### PR TITLE
feat(rome_service): auto generate configuration for rules

### LINKED ISSUE #2887: auto generate lint configuration for disabling rules

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

Closes #2887 

This PR implements a way to auto generate the configuration of the `rome.json` of the rules using the metadata defined in each rule. 

The configuration, in this PR, is now shaped in this way:

```json
{
  "linter": {
    "enabled": true,
    "globals": ["$"],
    "rules": {
        "js": {
            "noDeadCode": "off",
            "useSimplifiedLogicExpression": "warn"
        }
    }
  }
}
```

There's a new field called `globals`. We will use this field to store global references to the root scope of the semantic model cc @xunilrj 

There's a new field called `rules` which will contain the rules. The rules divided by groups. This all auto generate from the metadata, so it will change accordingly cc @leops .

Each rule is defined by a `RuleConfiguration`, where we can:
- turn off a rule
- make a rule to emit warnings
- make a rule to emit errors

A rule is **not serialized** if it's configured to emit diagnostics, because it's the default. This could change if we will use a recommendation system.

Consuming these defaults **is out of scope**, this will be implemented in other PRs. 

As for now, all the rules are enabled and their default is `RuleConfiguration::Error`, although since they are not consumed yet, it won't make a difference for now. 

The idea is to enable certain rules as "recommended", but this is also out of scope and it will be implemented in another PR.

I also fixed an error on the max diagnostics printer, the message was incorrectly emitted every  time.

## Test Plan

Because the configuration is not consume it yet, there's no practical test. I added for unknown field on the rule, which demonstrates that the configuration is correctly deserialized by `serde`.

### CODE REVIEW

File: crates/rome_service/src/configuration/linter/rules.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
+//! Generated file, do not edit by hand, see `xtask/codegen`
+
+use serde::{Deserialize, Serialize};
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Clone)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub enum RuleConfiguration {
+    Warn,
+    Error,
+    Off,
+}
```
Comment (MichaReiser) [inline crates/rome_service/src/configuration/linter/rules.rs:10]: Is it possible to move `RuleConfiguration` out of the generated file? It seems to me that its implementation doesn't depend on any generated code.

File: crates/rome_service/src/settings.rs [hunk 2]
```diff
@@ -55,16 +61,27 @@ impl Default for FormatSettings {
...
+/// Linter settings for the entire workspace
 #[derive(Debug)]
 pub struct LinterSettings {
     /// Enabled by default
     pub enabled: bool,
+
+    /// A list of global bindings that should be ignored by the analyzers
+    ///
+    /// If defined here, they should not emit diagnostics.
+    pub globals: Option<Vec<String>>,
```
Comment (MichaReiser) [inline crates/rome_service/src/settings.rs:73]: Is there a semantic difference between an empty vector and `None`? If not, then I would recommend removing the `Option` wrapping of the vector.

**Review by MichaReiser — COMMENTED** [review]
Looks good. I do have a few questions (and I like @leops point on the suggestions). 

* Does Rome-classic support changing the warning level of a diagnostic? 
* 'error' is likely not the right default for all lint rules, especially rules for which no sound static analysis is possible due to JS's dynamic nature. How do you intend rules that, by default, should emit warnings?

File: crates/rome_service/src/settings.rs (hunk 2 above)
Comment (ematipico) [reply crates/rome_service/src/settings.rs:73]: That should work I think, we can use `Vec::is_empty` to avoid the serialization

**Comment by leops** [comment]
> I would like to evaluate the PR with this approach. We are always in time for changing it.

Absolutely, this is just something I thought about while reviewing this and wanted to raise a possible discussion about, but it's not an issue for merging this PR

> > 'error' is likely not the right default for all lint rules, especially rules for which no sound static analysis is possible due to JS's dynamic nature. How do you intend rules that, by default, should emit warnings?
> 
> That's a nice question, and I don't have an answer yet. The thing is, for now all rules emit warnings or errors and we don't have a more finer way of doing that per rule, not ahead of time at least. Each rule emits a diagnostic and we can't know ahead of time which type, unless we execute them.

This is something we can revisit in more detail once this configuration actually gets integrated with the analyzer, but it would be possible to make the severity on `RuleDiagnostic` optional and let it be filled with the severity associated with the `RuleConfiguration` for that rule by default, unless the rule specified an explicit severity for the diagnostic.

An alternative / complement to this would be to consider the `RuleConfiguration` enum as setting a "maximum level" at which a certain rule can emit diagnostics:
- if a rule is set to `RuleConfiguration::Error` and emits a diagnostic with a severity of `Error`, `Warning` or `Note` the diagnostic is emitted as-is
- if a rule is set to `RuleConfiguration::Warn` and emits a diagnostic with a severity of `Error` or `Warning` it gets emitted as a warning in both case, a severity of `Note` is still emitted as-is
- if a rule is set to `RuleConfiguration::Off` no diagnostic gets emitted

This way if a rule is not guaranteed to be sound it could emit a warning, and the diagnostic will still be emitted as a warning even if the rule is set to `RuleConfiguration::Error` (this is potentially very counterintuitive though)

**Comment by ematipico** [comment]
> I agree on the philosophy of providing good defaults and avoiding any unnecessary options. The philosophy could even be extended to not have support for overriding the lint rule level by arguing that we pick the appropriate defaults.

True, this is a good point. This is something we wanted to do time ago: https://github.com/rome/tools/issues/1529

> I think it would be worth spending some time exploring configuration formats that support per-rule options to avoid a breaking change in the very near future OR at least go through the rules with planned and check if we can implement all of them without having per-rule option support.

This is part of #2741 , where we will evaluate rules that require a configuration and discuss them, and see how we can go from there. 

I don't mind breaking changes because we are still in `0.*`. It allows us to experiment, change things and whatnot without worrying too much. We can start the exploration after we release `0.8.0`.

**Comment by ematipico** [comment]
@leops do you have any idea why the workflow for the codegen is failing? Can't understand why

**Review by MichaReiser — APPROVED** [review]

**Comment by ematipico** [comment]
> > I don't mind breaking changes because we are still in 0.*. It allows us to experiment, change things and whatnot without worrying too much. We can start the exploration after we release 0.8.0.
> 
> That's true but comes at the risk that other things will be more important in the upcoming 0.9 and then 10.0 release and we then missed our chance. That's why I think it would be valuable to invest a couple of minutes to post a few options on how we could support rule-specific options. For example, one option that isn't a breaking change is to define the rule value as `string | { level: "warn" | "off", options: {} }` or maybe we find a format that avoids the need for a union type (what do linters of other languages use?)

I believe we can use a mixed approach using what clippy is doing, which by using a custom function to deserialize the rules. Clippy doesn't allow to turn off rules at configuration level though, you can only provide options to rules. Clippy uses macros to turn off rules in a global way.

I will spend some time and propose a solution

**Pull request merged**